### ISSUE TITLE
kubelet cannot execute runc

### ISSUE THREAD

**Issue by yanovich** [canonical/microk8s#2645]
Please run `microk8s inspect` and attach the generated tarball to this issue.

We appreciate your feedback. Thank you for using microk8s
[inspection-report-20211012_103831.tar.gz](https://github.com/ubuntu/microk8s/files/7327922/inspection-report-20211012_103831.tar.gz)
.

```
$ kubectl describe pod kube-registry-699c4fd457-skkvd
...
Events:
  Type     Reason         Age                   From     Message
  ----     ------         ----                  ----     -------
  Warning  FailedKillPod  17m                   kubelet  error killing pod: [failed to "KillContainer" for "registry" with KillContainerError: "rpc error: code = Unknown desc = failed to stop container \"d30b2fa7037f0abeb639fc89afffe1319ed57642bbc8431b99f851e7f792496c\": unknown error after kill: exec: \"runc\": executable file not found in $PATH: <nil>: unknown", failed to "KillPodSandbox" for "ff5457f3-73a4-4c8f-a9cc-569214da724a" with KillPodSandboxError: "rpc error: code = Unknown desc = failed to stop container \"d30b2fa7037f0abeb639fc89afffe1319ed57642bbc8431b99f851e7f792496c\": failed to kill container \"d30b2fa7037f0abeb639fc89afffe1319ed57642bbc8431b99f851e7f792496c\": unknown error after kill: exec: \"runc\": executable file not found in $PATH: <nil>: unknown"]
  Normal   Killing        9m31s (x13 over 17m)  kubelet  Stopping container registry
  Warning  FailedKillPod  2m34s (x21 over 16m)  kubelet  error killing pod: [failed to "KillContainer" for "registry" with KillContainerError: "rpc error: code = Unknown desc = failed to kill container \"d30b2fa7037f0abeb639fc89afffe1319ed57642bbc8431b99f851e7f792496c\": unknown error after kill: exec: \"runc\": executable file not found in $PATH: <nil>: unknown", failed to "KillPodSandbox" for "ff5457f3-73a4-4c8f-a9cc-569214da724a" with KillPodSandboxError: "rpc error: code = Unknown desc = failed to stop container \"d30b2fa7037f0abeb639fc89afffe1319ed57642bbc8431b99f851e7f792496c\": failed to kill container \"d30b2fa7037f0abeb639fc89afffe1319ed57642bbc8431b99f851e7f792496c\": unknown error after kill: exec: \"runc\": executable file not found in $PATH: <nil>: unknown"]
```

-- 
**Diagnostics results:**
microk8s has `KillMode=process` in `/etc/systemd/system/snap.microk8s.daemon-containerd.service`. On snap update containerd is restarted, but child containers are left running. This is done to allow updates without disturbing cluster payload. When a container survives 2 snap updates, the snap version that launched the container is deleted. Since each version is an independent package, a running container has no management tools after the second snap update.

The **solution needs** to place a symlink to `runc` in `/snap/bin/`and ensure that it is called from there. The real solution is to provide a deb package for microk8s and install `runc` to `/usr/bin/`. For now I use the simplest workaround which is to disable automatic snap updates by setting a fake proxy:
```
snap set system proxy.http="http://127.0.0.1:1111"
snap set system proxy.https="http://127.0.0.1:1111"
```
When I am ready to update:
```
snap unset system proxy.http
snap unset system proxy.https
snap refresh
snap set system proxy.http="http://127.0.0.1:1111"
snap set system proxy.https="http://127.0.0.1:1111"
```

**Comment by yanovich** [comment]
I am not sure, but it seems to be related to a snap update. Only one of five identical nodes was affected.

**Comment by yanovich** [comment]
It looks like I found the cause. This seems to happen if the pod is launched by a version which has been removed by snap as too old

**Comment by jhughes2112** [comment]
I have this happening to me right now.  I'm going to do a `kubectl delete --force pod/xxxx` and hope it doesn't leave some dangling resources.  Is there a resolution to this issue?  Do I need to go to a node and check the snaps?

**Comment by jhughes2112** [comment]
Two things:  First, when I force deleted the pod in kubectl, it did leave the container running on the node.  I could not reapply the YAML manifest to re-launch it, because it was re-fetching the image and I think there was a problem with the container interfering with that process, somehow.  Second, when I took the whole node off-line with `microk8s stop` then `microk8s start`, the rogue containers started running again!  There doesn't appear to be a way to stop a container, only kill it, and that refuses to work if the container is still running.  I have no idea how to use micrkok8s to actually stop a container.

Anyway, it seems in my case the rogue container stopped and disappeared after another 10 minutes or so.  No idea why it went away, but it did.

**Comment by AlexGustafsson** [comment]
This happens me too, but I'm really unsure why. It's happened before and again now, usually after having run the cluster without changes for a while.

```
(combined from similar events): Readiness probe errored: rpc error: code = Unknown desc = failed to exec in container: failed to start exec "1aa2dce570a2c55d1e539c16e36806278595cbc5b9f61f680a8db7201eeea830": OCI runtime exec failed: exec: "runc": executable file not found in $PATH: unknown
```

```
error killing pod: [failed to "KillContainer" for "minio" with KillContainerError: "rpc error: code = Unknown desc = failed to kill container \"2709967c67b83ce6449ab1fc1c6de363eb240ab5403c47fa629a526ac285b7f5\": unknown error after kill: exec: \"runc\": executable file not found in $PATH: <nil>: unknown", failed to "KillPodSandbox" for "9fb404ce-f537-4a86-bb57-37f5fef3f249" with KillPodSandboxError: "rpc error: code = Unknown desc = failed to stop container \"2709967c67b83ce6449ab1fc1c6de363eb240ab5403c47fa629a526ac285b7f5\": failed to kill container \"2709967c67b83ce6449ab1fc1c6de363eb240ab5403c47fa629a526ac285b7f5\": unknown error after kill: exec: \"runc\": executable file not found in $PATH: <nil>: unknown"]
```

**Comment by tpo** [comment]
Same here:
```
Nov 30 08:44:17 the-host microk8s.daemon-kubelite[2434872]: E1130 08:44:17.549750 2434872 kubelet.go:1789] [failed to "KillContainer" for "the-pod" with KillContainerError: "rpc error: code = Unknown desc = failed to stop container \"262419645f9736a3f4675b34ff88d41de78eca689a867f3430412af668177579\": unknown error after kill: exec: \"runc\": executable file not found in $PATH: <nil>: unknown", failed to "KillPodSandbox" for "1d0bffff-dbe2-47fe-a5b3-99a5b6b7e9c9" with KillPodSandboxError: "rpc error: code = Unknown desc = failed to stop container \"262419645f9736a3f4675b34ff88d41de78eca689a867f3430412af668177579\": failed to kill container \"262419645f9736a3f4675b34ff88d41de78eca689a867f3430412af668177579\": unknown error after kill: exec: \"runc\": executable file not found in $PATH: <nil>: unknown"]
```
not sure why microl8s has to repeat the same error:
* in one line
* multiple times
it *certainly* doesn't make the log easier to read and easier to debug

now let's see the `PATH` environment that that `microk8s.daemon-kubelite` has:

# cat /proc/2434872/environ | tr '\0' '\n' 
SNAP_REVISION=2695
SNAP_REAL_HOME=/root
SNAP_USER_COMMON=/root/snap/microk8s/common
SNAP_INSTANCE_KEY=
PWD=/var/snap/microk8s/2695
SNAP_CONTEXT=[EDITED]
LANG=en_US.UTF-8
SNAP_ARCH=amd64
SNAP_INSTANCE_NAME=microk8s
SNAP_USER_DATA=/root/snap/microk8s/2695
INVOCATION_ID=[EDITED]
SNAP_REEXEC=
SNAP=/snap/microk8s/2695
SNAP_COMMON=/var/snap/microk8s/common
SNAP_VERSION=v1.22.4
SHLVL=0
SNAP_LIBRARY_PATH=/var/lib/snapd/lib/gl:/var/lib/snapd/lib/gl32:/var/lib/snapd/void
SNAP_COOKIE=[EDITED]
SNAP_DATA=/var/snap/microk8s/2695
LD_LIBRARY_PATH=/var/lib/snapd/lib/gl:/var/lib/snapd/lib/gl32:/var/lib/snapd/void::/snap/microk8s/2695/lib:/snap/microk8s/2695/usr/lib:/snap/microk8s/2695/lib/x86_64-linux-gnu:/snap/microk8s/2695/usr/lib/x86_64-linux-gnu
XDG_RUNTIME_DIR=/var/snap/microk8s/common/run
SNAP_NAME=microk8s
JOURNAL_STREAM=9:64403828
PATH=/snap/microk8s/2695/usr/sbin:/snap/microk8s/2695/usr/bin:/snap/microk8s/2695/sbin:/snap/microk8s/2695/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin
root@microk8s-node-0:/proc/2434872# PATH=/snap/microk8s/2695/usr/sbin:/snap/microk8s/2695/usr/bin:/snap/microk8s/2695/sbin:/snap/microk8s/2695/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin
```

Let's see whether `runc` is available in `microk8s.daemon-kubelite`'s path:

```
# PATH=/snap/microk8s/2695/usr/sbin:/snap/microk8s/2695/usr/bin:/snap/microk8s/2695/sbin:/snap/microk8s/2695/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin
# command -v runc
/snap/microk8s/2695/bin/runc
```

it is. Let's try to execute it:

```
# runc --help
NAME:
   runc - Open Container Initiative runtime

[...and so on...]
```
So `runc` is in `microk8s.daemon-kubelite`'s PATH and is correctly executable. So why doesn't `microk8s.daemon-kubelite` find it? An interesting bit in the log entry is the "nil" it contains:

```
[...] unknown error after kill: exec: \"runc\": executable file not found in $PATH: <nil>: unknown [...]
```

Does that mean that `microk8s.daemon-kubelite`'s PATH isn't set **at all**? Why would that be? Any idea @yanovich @ktsakalozos @jhughes2112 @AlexGustafsson ?

@AlexGustafsson writes:

> Last time this happened I had to drain and restart each node, one by one.

That the point - I can not drain the node, because then the I get the `executable not in $PATH` problem and the node does *not* drain.

**Comment by yanovich** [comment]
microk8s has `KillMode=process` in `/etc/systemd/system/snap.microk8s.daemon-containerd.service`. On snap update containerd is restarted, but child containers are left running. This is done to allow updates without disturbing cluster payload. When a container survives 2 snap updates, the snap version that launched the container is deleted. Since each version is an independent package, a running container has no management tools after the second snap update.

The **solution needs** to place a symlink to `runc` in `/snap/bin/`and ensure that it is called from there. The real solution is to provide a deb package for microk8s and install `runc` to `/usr/bin/`. For now I use the simplest workaround which is to disable automatic snap updates by setting a fake proxy:
```
snap set system proxy.http="http://127.0.0.1:1111"
snap set system proxy.https="http://127.0.0.1:1111"
```
When I am ready to update:
```
snap unset system proxy.http
snap unset system proxy.https
snap refresh
snap set system proxy.http="http://127.0.0.1:1111"
snap set system proxy.https="http://127.0.0.1:1111"
```

**Comment by jhughes2112** [comment]
Great explanation--I think I understand why it's broken now.  Thanks!

By the way, I think you should be able to set the snap system to hold forever, and still manually run refreshes.
```
$ sudo snap refresh --time
timer: 00:00~24:00/4
last: today at 13:10 UTC
next: today at 18:23 UTC

$ sudo snap set system refresh.hold="2100-01-01T00:00:00Z"

$ sudo snap refresh --time
timer: 00:00~24:00/4
last: today at 13:10 UTC
hold: 2022-03-05
next: today at 18:23 UTC (but held)
```

**Comment by ktsakalozos** [comment]
@yanovich, all, thank you for the awesome debugging work you did in this issue. Let me try to re-iterate what the problem is. When containerd spawns a new container it does so through a containerd-shim. The containerd-shim is the one calling `runc` when needed therefore it needs to always find it in its PATH. containerd-shims inherit their PATH from containerd. Contrary to containerd the shims persist through snap refreshes. In a refresh containerd gets updated with the new location of `runc` but the shims do not. The fix (linked to this issue above) is to have the `/snap/microk8s/current/bin/runc` in the containerd's PATH so it gets inherited by the shims.

For the workaround while we this patch makes its way through the release process is to put a `runc` in the PATH of the containerd-shims. Here are a couple of ways:
  * `apt install runc` on the node,
  * create a symbolic lint towards the `runc` of the snap: `sudo ln -s  /snap/microk8s/current/bin/runc /snap/bin/runc` . Other locations should also work, eg `sudo ln -s  /snap/microk8s/current/bin/runc /usr/bin/runc`.

Another workaround is to reboot the node.

Apologies for the inconvenience we may have caused. We are working to fix this.

**Comment by tpo** [comment]
@ktsakalozos would it be realistic to clean up the logging too so that the same thing isn't logged N times into the same log line (as seen in the first line [here](https://github.com/ubuntu/microk8s/issues/2645#issuecomment-982392308))?

**Comment by ktsakalozos** [comment]
@tpo there is the `-v` [1] option you can use on `/var/snap/microk8s/current/args/kubelet` to reduce the log verbosity level. Would that work for you?

[1] https://kubernetes.io/docs/concepts/cluster-administration/system-logs/

**Comment by benben** [comment]
I can confirm that the workaround by @ktsakalozos works.

**Comment by berlincount** [comment]
seems like this is troublesome in `1.23.2` still as well.

**Comment by ktsakalozos** [comment]
Hi @berlincount, this issue has been addressed in the PR linked to this issue.

On new deployments the path to runc includes the `/snap/microk8s/current/bin/` that holds `runc`. For old deployments we linked `/snap/microk8s/current/bin/runc` to  `/usr/sbin/runc`. I wonder how you get this problem. Is it a long lasting deployment? Maybe you refreshed to 1.23 before these fixes where applied to 1.22?
